File: src/ustplayer/core/settings_store.py
# settings_store.py — Settings.json 文件存取
"""设置持久化服务：新版 Settings.json（结构为「分组 → 键值」字典），
旧版 Settings.ini 在首次启动时自动迁移（迁移成功后删除旧文件）。

仅负责文件 IO 与迁移，不含任何设置业务逻辑。
"""

import configparser
import json
import os

from ustplayer.core.contracts import ensure_writable_dir, resolve_program_root
from ustplayer.core.log import logger


class SettingsStore:
    """设置存取服务（JSON）。"""

    FILE_NAME = "Settings.json"
    LEGACY_FILE_NAME = "Settings.ini"

    def __init__(self):
        self.settings_path = self._resolve_settings_path()

    @staticmethod
    def _resolve_settings_path() -> str:
        """解析 Settings.json 路径：程序根目录优先，实际不可写时回退用户数据目录。"""
        root = resolve_program_root()
        if ensure_writable_dir(root):
            return os.path.join(root, SettingsStore.FILE_NAME)
        fallback_dir = os.path.join(
            os.environ.get("LOCALAPPDATA", os.path.expanduser("~")), "ustPlayer"
        )
        try:
            os.makedirs(fallback_dir, exist_ok=True)
        except OSError:
            pass
        return os.path.join(fallback_dir, SettingsStore.FILE_NAME)

    def load(self) -> dict:
        """读取设置；JSON 不存在时尝试从旧版 Settings.ini 迁移；都没有则返回空配置。"""
        if os.path.exists(self.settings_path):
            try:
                # utf-8-sig 兼容 Windows 记事本等编辑器写入的 UTF-8 BOM
                with open(self.settings_path, "r", encoding="utf-8-sig") as f:
                    data = json.load(f)
                if isinstance(data, dict):
                    return data
                logger.warning(f"设置文件格式异常，按空配置处理: {self.settings_path}")
                return {}
            except Exception as e:
                logger.exception(f"读取设置文件失败，按空配置处理: {e}")
                return {}

        legacy = self._legacy_path()
        if legacy and os.path.exists(legacy):
            return self._migrate_legacy(legacy)
        return {}

    @staticmethod
    def _fallback_path() -> str:
        """用户数据目录回退路径（与 _resolve_settings_path 保持一致）。"""
        fallback_dir = os.path.join(
            os.environ.get("LOCALAPPDATA", os.path.expanduser("~")), "ustPlayer"
        )
        return os.path.join(fallback_dir, SettingsStore.FILE_NAME)

    def save(self, config: dict):
        """把配置写入 Settings.json（临时文件 + 原子替换，避免写一半损坏配置）。

        首选路径的目录虽可写、但目标文件本身可能只读 / 被 ACL 拒绝替换
        （Windows 常见），此时自动改写到 %LOCALAPPDATA%\\ustPlayer 并切换
        后续使用的 settings_path，避免设置从此静默丢失。
        """
        if self._try_save(self.settings_path, config):
            return

        fallback = self._fallback_path()
        logger.warning(f"首选设置文件不可替换，回退到: {fallback}")
        if self._try_save(fallback, config):
            self.settings_path = fallback

    @staticmethod
    def _try_save(target_path: str, config: dict) -> bool:
        """写临时文件并原子替换到 target_path；任何一步失败都清理临时文件。"""
        tmp_path = target_path + ".tmp"
        try:
            os.makedirs(os.path.dirname(target_path), exist_ok=True)
            with open(tmp_path, "w", encoding="utf-8") as f:
                json.dump(config, f, ensure_ascii=False, indent=2)
            os.replace(tmp_path, target_path)
            return True
        except OSError as e:
            logger.warning(f"写入设置失败: {target_path} ({e})")
            return False
        finally:
            try:
                if os.path.exists(tmp_path):
                    os.remove(tmp_path)
            except OSError:
                pass

    def _legacy_path(self) -> str:
        """旧版 Settings.ini 路径（与新版同目录逻辑）。"""
        root = resolve_program_root()
        if ensure_writable_dir(root):
            return os.path.join(root, self.LEGACY_FILE_NAME)
        fallback_dir = os.path.join(
            os.environ.get("LOCALAPPDATA", os.path.expanduser("~")), "ustPlayer"
        )
        return os.path.join(fallback_dir, self.LEGACY_FILE_NAME)

    def _migrate_legacy(self, legacy_path: str) -> dict:
        """旧版 Settings.ini → 分组字典，写为新版 Settings.json 并删除旧文件。

        旧版值允许包含未转义的 %（如项目名 100% Pure），因此必须关闭
        ConfigParser 的插值，否则迁移会抛 InterpolationSyntaxError。
        """
        parser = configparser.ConfigParser(interpolation=None)
        try:
            parser.read(legacy_path, encoding="utf-8-sig")
        except Exception as e:
            logger.exception(f"读取旧版设置文件失败: {e}")
            return {}
        try:
            config = {section: dict(parser[section]) for section in parser.sections()}
        except Exception as e:
            logger.exception(f"解析旧版设置文件失败: {e}")
            return {}
        try:
            self.save(config)
            os.remove(legacy_path)
            logger.info(f"已从旧版 {self.LEGACY_FILE_NAME} 迁移设置至 {self.FILE_NAME}")
        except Exception as e:
            logger.exception(f"旧版设置迁移失败（旧文件保留）: {e}")
        return config

File: ustPlayer/Settings/SettingsStore.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;

using UstPlayer.Diagnostics;

namespace UstPlayer.Settings;

/// <summary>
/// <c>Settings.json</c> 的文件 I/O — 从 1.1.x <c>core/settings_store.py</c> 的
/// <c>SettingsStore</c> 移植。
/// </summary>
/// <remarks>
/// <para>
/// 只负责文件读写与旧格式迁移，不含任何设置业务逻辑。
/// </para>
/// <para>
/// 三处与 1.1.x 逐字对齐的行为：
/// </para>
/// <list type="number">
///   <item><b>原子写</b>：先写 <c>*.tmp</c> 再整体替换。崩溃或断电不会留下半截配置。</item>
///   <item><b>首选路径不可替换时改写回退路径</b>：首选目录可写并不代表目标文件可替换
///   （只读文件 / ACL 拒绝替换，Windows 常见），此时改写到用户数据目录**并切换后续使用的路径**，
///   避免设置从此静默丢失。</item>
///   <item><b>旧版 <c>Settings.ini</c> 自动迁移</b>：首次运行时把 ini 转成 JSON 并删除旧文件。
///   值里可能含裸 <c>%</c>（如工程名「100% Pure」），因此解析时**不做插值**——
///   1.1.x 正是因为没有关掉 ConfigParser 的插值而崩溃过。</item>
/// </list>
/// </remarks>
internal sealed class SettingsStore
{
    /// <summary>设置文件名。</summary>
    internal const string FileName = "Settings.json";

    /// <summary>旧版设置文件名。</summary>
    internal const string LegacyFileName = "Settings.ini";

    /// <summary>崩溃前写入的临时文件名（保留以兼容 1.1.x 遗留文件）。</summary>
    private const string TempSuffix = ".tmp";

    /// <summary>
    /// JSON 序列化选项：<b>与 1.1.x 的输出逐字对齐</b>。
    /// </summary>
    /// <remarks>
    /// Python 侧是 <c>json.dump(config, f, ensure_ascii=False, indent=2)</c>：
    /// 两空格缩进、中文不转义。这里用同一组设置，使两个版本写出的文件在格式上一致
    /// （便于用户用 diff 核对，也避免无谓的整文件重写）。
    /// </remarks>
    private static readonly JsonSerializerOptions SerializerOptions = new()
    {
        WriteIndented = true,
        IndentSize = 2,
        Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
    };

    private string _settingsPath;

    /// <summary>创建存储并解析设置文件路径（程序目录优先，不可写时回退用户数据目录）。</summary>
    internal SettingsStore()
        : this(settingsPath: null)
    {
    }

    /// <summary>创建存储并指定设置文件路径。</summary>
    /// <param name="settingsPath">
    /// 设置文件路径；传 <see langword="null"/> 时按默认策略解析
    /// （程序目录优先，不可写时回退用户数据目录）。测试与将来的便携模式都走这里。
    /// </param>
    internal SettingsStore(string? settingsPath)
    {
        _settingsPath = string.IsNullOrWhiteSpace(settingsPath)
            ? ResolveSettingsPath()
            : settingsPath;
    }

    /// <summary>当前使用的设置文件绝对路径（回退后会切换）。</summary>
    internal string SettingsPath => _settingsPath;

    /// <summary>
    /// 读取设置。
    /// </summary>
    /// <returns>配置；文件不存在且无旧格式可迁移时返回空配置。</returns>
    /// <remarks>
    /// 文件损坏（非法 JSON）时返回空配置并记录日志，不抛异常——
    /// 与 1.1.x 一致：坏掉的配置文件不该让程序无法启动。
    /// </remarks>
    internal SettingsConfig Load()
    {
        if (File.Exists(_settingsPath))
        {
            return LoadJsonFile(_settingsPath);
        }

        var legacyPath = ResolveLegacyPath();
        if (legacyPath is not null && File.Exists(legacyPath))
        {
            return MigrateLegacy(legacyPath);
        }

        return SettingsConfig.Empty();
    }

    /// <summary>
    /// 保存设置（原子写；首选路径不可替换时改写回退路径）。
    /// </summary>
    /// <param name="config">配置。</param>
    internal void Save(SettingsConfig config)
    {
        ArgumentNullException.ThrowIfNull(config);

        if (TrySave(_settingsPath, config))
        {
            return;
        }

        var fallback = ResolveFallbackPath();
        AppLogger.Warning($"首选设置文件不可替换，回退到：{fallback}");

        if (TrySave(fallback, config))
        {
            _settingsPath = fallback;
        }
    }

    /// <summary>
    /// 读取 JSON 设置文件。
    /// </summary>
    /// <param name="path">文件路径。</param>
    /// <returns>配置。</returns>
    private static SettingsConfig LoadJsonFile(string path)
    {
        try
        {
            var text = File.ReadAllText(path, Encoding.UTF8);

            // 兼容记事本等编辑器写入的 UTF-8 BOM：BOM 会让 JSON 解析器直接报错
            var node = JsonNode.Parse(text.TrimStart('\uFEFF'));

            if (node is JsonObject root)
            {
                return new SettingsConfig(root);
            }

            AppLogger.Warning($"设置文件格式异常（顶层不是对象），按空配置处理：{path}");
            return SettingsConfig.Empty();
        }
        catch (Exception exception)
        {
            AppLogger.Error($"读取设置文件失败，按空配置处理：{path}", exception);
            return SettingsConfig.Empty();
        }
    }

    /// <summary>原子写入：先写临时文件再替换。</summary>
    /// <param name="targetPath">目标路径。</param>
    /// <param name="config">配置。</param>
    /// <returns>成功返回 <see langword="true"/>。</returns>
    private static bool TrySave(string targetPath, SettingsConfig config)
    {
        var tempPath = targetPath + TempSuffix;

        try
        {
            var directory = Path.GetDirectoryName(targetPath);
            if (!string.IsNullOrEmpty(directory))
            {
                Directory.CreateDirectory(directory);
            }

            // UTF-8 无 BOM：与 1.1.x 的 open(..., encoding="utf-8") 一致
            var json = config.Root.ToJsonString(SerializerOptions);
            File.WriteAllText(tempPath, json, new UTF8Encoding(encoderShouldEmitUTF8Identifier: false));

            File.Move(tempPath, targetPath, overwrite: true);
            return true;
        }
        catch (Exception exception)
        {
            AppLogger.Warning($"写入设置失败：{targetPath}（{exception.Message}）");
            return false;
        }
        finally
        {
            TryDelete(tempPath);
        }
    }

    /// <summary>
    /// 解析设置文件路径：程序目录优先，实际不可写时回退用户数据目录。
    /// </summary>
    /// <returns>设置文件路径。</returns>
    private static string ResolveSettingsPath() =>
        Path.Combine(ProgramPaths.SettingsDirectory(), FileName);

    /// <summary>用户数据目录下的回退路径。</summary>
    /// <returns>设置文件路径。</returns>
    private static string ResolveFallbackPath() =>
        Path.Combine(ProgramPaths.UserDataDirectory, FileName);

    /// <summary>旧版 <c>Settings.ini</c> 的路径（与设置文件同目录逻辑）。</summary>
    /// <returns>路径；目录不可判定时返回 <see langword="null"/>。</returns>
    private static string? ResolveLegacyPath()
    {
        var directory = ProgramPaths.SettingsDirectory();
        return string.IsNullOrEmpty(directory) ? null : Path.Combine(directory, LegacyFileName);
    }

    /// <summary>
    /// 把旧版 <c>Settings.ini</c> 迁移为 JSON，成功后删除旧文件。
    /// </summary>
    /// <param name="legacyPath">旧文件路径。</param>
    /// <returns>迁移得到的配置。</returns>
    private static SettingsConfig MigrateLegacy(string legacyPath)
    {
        var root = new JsonObject();

        try
        {
            foreach (var (section, values) in ParseIni(legacyPath))
            {
                var group = new JsonObject();
                foreach (var (key, value) in values)
                {
                    group[key] = value;
                }

                root[section] = group;
            }
        }
        catch (Exception exception)
        {
            AppLogger.Error($"解析旧版设置文件失败：{legacyPath}", exception);
            return SettingsConfig.Empty();
        }

        var config = new SettingsConfig(root);

        try
        {
            var store = new SettingsStore();
            store.Save(config);
            File.Delete(legacyPath);
            AppLogger.Info($"已从旧版 {LegacyFileName} 迁移设置至 {FileName}");
        }
        catch (Exception exception)
        {
            AppLogger.Error("旧版设置迁移失败（旧文件保留）", exception);
        }

        return config;
    }

    /// <summary>
    /// 解析 INI 文本（段 → 键值）。
    /// </summary>
    /// <param name="path">文件路径。</param>
    /// <returns>段与键值。</returns>
    /// <remarks>
    /// <b>不做任何插值</b>：值里的 <c>%</c> 按字面处理。1.1.x 曾因 ConfigParser 默认插值
    /// 在含 <c>%</c> 的值上抛 <c>InterpolationSyntaxError</c> 导致迁移崩溃。
    /// 只按第一个 <c>=</c> 切分，值里的 <c>=</c> 原样保留。
    /// </remarks>
    private static IEnumerable<(string Section, List<(string Key, string Value)> Values)> ParseIni(
        string path)
    {
        var result = new List<(string, List<(string, string)>)>();
        List<(string Key, string Value)>? current = null;

        foreach (var rawLine in File.ReadLines(path, Encoding.UTF8))
        {
            var line = rawLine.Trim().TrimStart('\uFEFF');

            if (line.Length == 0 || line[0] is ';' or '#')
            {
                continue;
            }

            if (line.StartsWith('[') && line.EndsWith(']'))
            {
                current = [];
                result.Add((line[1..^1].Trim(), current));
                continue;
            }

            var separator = line.IndexOf('=');
            if (separator < 0)
            {
                continue;
            }

            current?.Add((line[..separator].Trim(), line[(separator + 1)..].Trim()));
        }

        return result;
    }

    private static void TryDelete(string path)
    {
        try
        {
            if (File.Exists(path))
            {
                File.Delete(path);
            }
        }
        catch (Exception)
        {
            // 临时文件清理失败无关紧要
        }
    }
}
